src/bif/erlang.rs

```diff
@@ -0,0 +1,519 @@
+pub fn bif_erlang_make_tuple_2(_vm: &vm::Machine, process: &RcProcess, args: &[Term]) -> BifResult {
+    let num = match args[0].into_number() {
+        Ok(value::Num::Integer(i)) if ! i < 0 => i,
+        _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    let heap = &process.context_mut().heap;
+    let tuple = value::tuple(&heap, num as u32);
+    for i in 0..num {
+        unsafe {
+            std::ptr::write(&mut tuple[i as usize], args[1].clone());
+        }
+    }
+    Ok(Term::from(tuple))
+}
+
+pub fn bif_erlang_make_tuple_3(_vm: &vm::Machine, process: &RcProcess, args: &[Term]) -> BifResult {
+    let num = match args[0].into_number() {
+       Ok(value::Num::Integer(i)) if ! i < 0 => i,
+       _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    let heap = &process.context_mut().heap;
+    let tuple = value::tuple(&heap, num as u32);
+    for i in 0..num {
+        unsafe {
+            std::ptr::write(&mut tuple[i as usize], args[1].clone());
+        }
+    }
+    if !args[2].is_list() {
+        return Err(Exception::new(Reason::EXC_BADARG));
+    }
+    let init: &value::Cons = match args[2].try_into() {
+        Ok(cons) => cons,
+        _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    for item in init.iter() {
+        let t: &Tuple = match item.try_into() {
+            Ok(tuple) => tuple, // FIXME do the len checking here
+            _ => return Err(Exception::new(Reason::EXC_BADARG))
+        };
+        if t.len != 2 {
+            return Err(Exception::new(Reason::EXC_BADARG))
+        }
+        let n = match t[0].into_number() {
+            Ok(value::Num::Integer(i)) if ! i < 1 && i - 1 < num => i - 1,
+            _ => return Err(Exception::new(Reason::EXC_BADARG))
+        };
+        unsafe {
+            std::ptr::write(&mut tuple[n as usize], t[1]);
+        }
+    }
+    Ok(Term::from(tuple))
+}
+
+pub fn bif_erlang_append_element_2(_vm: &vm::Machine, process: &RcProcess, args: &[Term]) -> BifResult {
+    if !args[0].is_tuple() {
+        return Err(Exception::new(Reason::EXC_BADARG));
+    }
+    let t: &Tuple = match args[0].try_into() {
+        Ok(tuple) => tuple,
+        _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    let heap = &process.context_mut().heap;
+    let mut new_tuple = value::tuple(&heap, (t.len() + 1) as u32);
+    unsafe {
+        new_tuple[..t.len()].copy_from_slice(&t[..]);
+        std::ptr::write(&mut new_tuple[t.len()], args[1]);
+    }
+    Ok(Term::from(new_tuple))
+}
+
+pub fn bif_erlang_setelement_3(_vm: &vm::Machine, process: &RcProcess, args: &[Term]) -> BifResult {
+    let number = match args[0].into_number() {
+        Ok(value::Num::Integer(i)) if ! i < 1 => i - 1,
+        _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    let t: &Tuple = match args[1].try_into() {
+        Ok(tuple) => tuple,
+        _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    if !(number <= t.len() as i32) {
```

Comment:
This is an off-by-one, since a `len = 2`, `number = 2` would mean we'd be writing into element 3 (0,1,2). You want `number < t.len()`

Reply:
I have tested the erlang:setelement with erl.
If we call ```erlang:setelement(2, {1, 2}, 3).```, we will get ```{1,3}``` instead of ```BADARG```.
So it is correct if ```len = 2```, ```number = 2```. The program will set the second element instead of the third one.

Reply:
Ah actually, the problem is when you have `len = 2`, `number = 3`, because your code will first do `number -= 1 => 2`, but then `<= t.len()` will allow this code through (indexing 2 out of [0,1]) and panic. I modified the test to show the crash here: https://github.com/archseer/enigma/pull/30/commits/605846eb93fd1ba0c9bdbd0a10d6fe21344c9090 then fixed it